### PR TITLE
Don't rely on device name provided by Cinder

### PR DESCRIPTION
**Pull request by kiall**
See issue #33128

We can't rely on the device name provided by Cinder, and thus must perform
detection based on the drive serial number (aka It's cinder ID) on the
kubelet itself.

This patch re-works the cinder volume attacher to ignore the supplied
deviceName, and instead defer to the pre-existing GetDevicePath method to
discover the device path based on it's serial number and /dev/disk/by-id
mapping.

This new behavior is controller by a config option, as falling back
to the cinder value when we can't discover a device would risk devices
not showing up, falling back to cinder's guess, and detecting the wrong
disk as attached.

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/33270)

### CODE REVIEW

**Comment by kiall** [comment]
cc / @anguslees @kubernetes/sig-openstack

**Comment by kiall** [comment]
So - Commit #2 (removal of Remove unused WaitForDetach) can actually be updated to handle this bug, now that a config option is present.

It would require updating to pass in the volume spec, so we can find the Cinder ID. Though, it still seems prudent to remove this unused code.

**Comment by kiall** [comment]
Rebased, anyone able to review? :)

**Review by anguslees — APPROVED** [review]
/lgtm

**Comment by anguslees** [comment]
@k8s-bot ok to test

**Comment by stuart-warren** [comment]
@kiall
>  cannot use testcase (type *testcase) as type CinderProvider in field value:
> 	*testcase does not implement CinderProvider (missing ShouldTrustDevicePath method)

**Comment by rootfs** [comment]
This PR mixes several issues. Could you split them, one for Cinder and one for WaitForDetach refactoring.

@kubernetes/sig-storage

**Comment by kiall** [comment]
@rootfs I can split that commit to a separate pull request, but GitHub will keep it here too as it doesn't have any dependencies concept (spoiled by gerrit over in OpenStack ;))

It's not possible to update the unused method to account for this fix, not with other unnecessary major changes. Hence, once I discovered it was unused, I removed it to proceed with this fix.

**Comment by rootfs** [comment]
@kiall yes multiple PRs are better

**Comment by kiall** [comment]
@rootfs done

File: pkg/cloudprovider/providers/openstack/openstack_volumes.go [hunk 1]
```diff
@@ -169,16 +169,25 @@ func (os *OpenStack) CreateVolume(name string, size int, vtype, availability str
 
 // GetDevicePath returns the path of an attached block storage volume, specified by its id.
 func (os *OpenStack) GetDevicePath(diskId string) string {
+	// Build a list of candidate device paths
+	candidateDeviceNodes := []string{
+		// KVM
```
Comment (rootfs) [inline pkg/cloudprovider/providers/openstack/openstack_volumes.go]: is there any doc on the device path format?

File: pkg/volume/cinder/attacher.go [hunk 2]
```diff
@@ -110,17 +110,14 @@ func (attacher *cinderDiskAttacher) Attach(spec *volume.Spec, nodeName types.Nod
...
 func (attacher *cinderDiskAttacher) WaitForAttach(spec *volume.Spec, devicePath string, timeout time.Duration) (string, error) {
+	// NOTE: devicePath is is path as reported by Cinder, which may be incorrect and should not be used. See Issue #33128
 	volumeSource, _, err := getVolumeSource(spec)
 	if err != nil {
 		return "", err
 	}
 
 	volumeID := volumeSource.VolumeID
 
-	if devicePath == "" {
```
Comment (rootfs) [inline pkg/volume/cinder/attacher.go]: why removing this? if devicePath is empty and `ShouldTrustDevicePath()` is true, then we end up with an empty device path to detect.

File: pkg/volume/cinder/attacher.go [hunk 3]
```diff
@@ -132,13 +129,17 @@ func (attacher *cinderDiskAttacher) WaitForAttach(spec *volume.Spec, devicePath
...
 			exists, err := volumeutil.PathExists(devicePath)
 			if exists && err == nil {
-				glog.Infof("Successfully found attached Cinder disk %q.", volumeID)
+				glog.Infof("Successfully found attached Cinder disk %q at %v.", volumeID, devicePath)
 				return devicePath, nil
 			} else {
-				//Log error, if any, and continue checking periodically
-				glog.Errorf("Error Stat Cinder disk (%q) is attached: %v", volumeID, err)
+				// Log an error, and continue checking periodically
+				glog.Errorf("Error could not find attached Cinder disk %q: %v", volumeID, err)
```
Comment (rootfs) [inline pkg/volume/cinder/attacher.go]: Error: could not find

**Comment by rootfs** [comment]
@kiall can you add a unit test? overall looks good to me

**Comment by kiall** [comment]
Will rebase, address comments, and add a test once I get back from the OpenStack Summit.

File: pkg/volume/cinder/attacher.go (hunk 2 above)
Comment (kiall) [inline pkg/volume/cinder/attacher.go]: Done

File: pkg/volume/cinder/attacher.go [hunk 4]
```diff
@@ -158,6 +159,22 @@ func (attacher *cinderDiskAttacher) GetDeviceMountPath(
 
 // FIXME: this method can be further pruned.
 func (attacher *cinderDiskAttacher) MountDevice(spec *volume.Spec, devicePath string, deviceMountPath string) error {
+	// NOTE: devicePath is is path as reported by Cinder, which may be incorrect and should not be used. See Issue #33128
```
Comment (kiall) [inline pkg/volume/cinder/attacher.go]: Ah, I see. Changes here have been reverted

**Comment by kiall** [comment]
I still need to re-test this against a real OpenStack deployment, but should be good to go otherwise.

**Comment by kiall** [comment]
Review status: 0 of 7 files reviewed at latest revision, 3 unresolved discussions.

---

*[pkg/volume/cinder/attacher.go, line 162 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/33270#-KSGt0eFYpG1epOMaVaa-r2-162:-KVdnr5t8MaPyWi1EDZE:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/0c9f7bda910b2bbb0da2480ef88a340f07887e56/pkg/volume/cinder/attacher.go#L162)):*
<details><summary><i>Previously, kiall (Kiall Mac Innes) wrote…</i></summary>
> Ah, I see. Changes here have been reverted
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/33270)*

**Comment by kiall** [comment]
I have no idea how to resolve the 3 reviewable discussions left!

**Comment by dims** [comment]
Looks good to me 👍

**Comment by dims** [comment]
@thockin @pmorie : can you please add milestone 1.5 for this one? and switch the release-note-label-needed tag to release-note-none please? oh please lgtm as well :)

**Comment by rootfs** [comment]
LGTM
adding @jsafrane @childsb

**Comment by jsafrane** [comment]
both @anguslees and @rootfs are fine with this PR, I am approving it, however I expect there will be subsequent PRs that update deployment scripts in `cluster/openstack-heat` and `cluster/rackspace` that add support for the new cloud.conf option + a PR for associated documentation in http://kubernetes.io/docs/getting-started-guides/openstack-heat/

**Comment by gnufied** [comment]
`isDeviceOpened` check similarly should use correct device path rather than relying on value present in node.Status blindly. This causes UnmountDevice to fail sometimes. cc @msau42

**Pull request merged**